### PR TITLE
SLS-595 Initial code to drain the ingest tables

### PR DESCRIPTION
**Pull request by quchenhao**
This is an initial implementation that hasn't been used and tested. Just try to get some feedback how it looks.

### CODE REVIEW

**Comment by quchenhao** [comment]
> _ No description provided. _

File: src/conn/conn_layered.c [hunk 1]
```diff
@@ -1276,3 +1276,172 @@ __wt_disagg_advance_checkpoint(WT_SESSION_IMPL *session, bool ckpt_success)
...
+            else {
+                prev_upd = upd;
+                upds = upd;
+            }
+            upd = NULL;
+        }
+    }
+
+err:
+    if (tombstone != NULL)
```
Comment (korteland) [inline src/conn/conn_layered.c:1410]: Close cursors?

File: src/conn/conn_layered.c [hunk 2]
```diff
@@ -1276,3 +1276,172 @@ __wt_disagg_advance_checkpoint(WT_SESSION_IMPL *session, bool ckpt_success)
...
+    WT_ERR(__wt_scr_alloc(session, 0, &key));
+    WT_ERR(__wt_scr_alloc(session, 0, &tmp_key));
+    WT_ERR(__wt_scr_alloc(session, 0, &value));
+
+    for (;;) {
+        tombstone = upd = NULL;
+        WT_ERR_NOTFOUND_OK(version_cursor->next(version_cursor), true);
+        if (ret == WT_NOTFOUND) {
+            if (key->size > 0 && upds != NULL)
+                WT_ERR(__layered_move_updates(cbt, key, upds));
```
Comment (korteland) [inline src/conn/conn_layered.c:1337]: I'm finding it hard to follow this part of the logic. Why do we have the `__layered_move_updates` call here? I suspect that `upds` can be non-NULL after the first iteration of the loop, but this call is after `cursor->next` so it looks like they go to the wrong key. But `__layered_move_updates` does a search for the key that was presumably set further down.

Why couldn't the control flow here go more something like this:
1. `cursor->next`
2. Potentially set `upds`
3. `__layered_move_updates`

Then that also removes the search call in `__layered_move_updates`.

**Review by korteland — COMMENTED** [review]
At a high level - wasn't the hard problem that we needed to find a way to avoid doing checkpoints while this is happening? I don't think that's addressed here, is it out of scope for this PR?

File: src/conn/conn_layered.c (hunk 2 above)
Comment (quchenhao) [reply src/conn/conn_layered.c:1337]: The cursor next call is on the version cursor. You cannot avoid doing a search on the stable cursor to apply the updates. Version cursor returns the values from newest to oldest for a single key so we need to build a chain of updates first. Then when we move to the next key or we have no key, we move the entire chain to the stable cursor in one go.

File: src/conn/conn_layered.c (hunk 1 above)
Comment (korteland) [reply src/conn/conn_layered.c:1410]: That's only the stable cursor though, the version cursor is something we open.

Comment (quchenhao) [reply src/conn/conn_layered.c:1410]: oh, good catch.

File: src/conn/conn_layered.c (hunk 2 above)
Comment (korteland) [reply src/conn/conn_layered.c:1337]: Ah, OK. Yeah I get this code now. Thanks for explaining it.

**Comment by quchenhao** [comment]
> At a high level - wasn't the hard problem that we needed to find a way to avoid doing checkpoints while this is happening? I don't think that's addressed here, is it out of scope for this PR?

This is out of scope. I think this code will be code during the reconfigure call before we set ourself to the leader. So no checkpoint can happen before we finish draining.

**Review by korteland — COMMENTED** [review]
Overall - I think the idea will work. I'm not sure about the `reconfig` call blocking, since that may not be where the server wants to block. I don't know where else they'd do it though.

Reconfig also takes the checkpoint lock, which is useful, but does that lock also stop concurrent reads and writes? If not, I'd be worried about how confident we are at the end of the loop that we really are done.

**Comment by quchenhao** [comment]
> Overall - I think the idea will work. I'm not sure about the `reconfig` call blocking, since that may not be where the server wants to block. I don't know where else they'd do it though.
> 
> Reconfig also takes the checkpoint lock, which is useful, but does that lock also stop concurrent reads and writes? If not, I'd be worried about how confident we are at the end of the loop that we really are done.



> Overall - I think the idea will work. I'm not sure about the `reconfig` call blocking, since that may not be where the server wants to block. I don't know where else they'd do it though.
> 
> Reconfig also takes the checkpoint lock, which is useful, but does that lock also stop concurrent reads and writes? If not, I'd be worried about how confident we are at the end of the loop that we really are done.

Checkpoint lock doesn't stop concurrent reads and writes. What do you mean that we are really done?

**Comment by korteland** [comment]
> Checkpoint lock doesn't stop concurrent reads and writes. What do you mean that we are really done?

My concern was about what happens if someone writes a new key/value while this loop is running. Thinking about it a bit more, I suspect we wouldn't allow the write since we're not finished reconfiguring as primary while this code is running. Is that correct?

**Comment by quchenhao** [comment]
> > Checkpoint lock doesn't stop concurrent reads and writes. What do you mean that we are really done?
> 
> My concern was about what happens if someone writes a new key/value while this loop is running. Thinking about it a bit more, I suspect we wouldn't allow the write since we're not finished reconfiguring as primary while this code is running. Is that correct?

Yes, that's the initial plan. If we want to allow concurrent writes, then we need to walk the trees multiple times. In the end, we will stop accepting writes and have a final round of walk before we can accept writes again.

**Comment by korteland** [comment]
> In the end, we will stop accepting writes and have a final round of walk before we can accept writes again.

That sounds like a good starting point. Thanks for answering my questions!

**Review by korteland — APPROVED** [review]

File: src/conn/conn_layered.c [hunk 3]
```diff
@@ -1276,3 +1276,173 @@ __wt_disagg_advance_checkpoint(WT_SESSION_IMPL *session, bool ckpt_success)
...
```
Comment (pmacko86) [inline src/conn/conn_layered.c:1337]: Should this also set `upds = NULL`, just like below?

Comment (quchenhao) [reply src/conn/conn_layered.c:1337]: yes, good catch

**Review by pmacko86 — APPROVED** [review]
LGTM. I have only cursory understanding of update chains, so I couldn't review this in too much depth. But overall it looks good. Thanks!
